### PR TITLE
[FSTORE-849] Support Spine Dataframes for batch data and training datasets

### PR DESCRIPTION
**Pull request by moritzmeister**
- add spine group creation
- add spine engine
- fix
- move features property to base class
- small fix
- cache df
- fix
- get or create and alias fix
- fix
- use spine fs
- pass dataframe down
- print
- dirty type check to avoid circluar dimport
- fix dirty hack
- fix
- pass down spine
- train_test_split with spine
- pass down spine
- change get or create error code
- allow passing spine group
- raise useful exception
- add docs
- fix unit tests

This PR adds/fixes/changes...
- please summarize your changes to the code 
- and make sure to include all changes to user-facing APIs

JIRA Issue: -

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [ ] Unit Tests
- [ ] Integration Tests
- [ ] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/feature_store.py [hunk 1]
```diff
@@ -903,6 +903,146 @@ def create_external_feature_group(
...
+        for a small set of primary key values. To do this, you can pass the spine group
+        instead of a dataframe. Just make sure it contains the needed primary key, event time and
+        label column.
+
+        ```python
+        feature_view.get_batch_data(spine=spine_group)
+        ```
+
+        # Arguments
+            name: Name of the external feature group to create.
```
Comment (jimdowling) [inline python/hsfs/feature_store.py:994]: Name of the spine group to create.

File: python/hsfs/feature_group.py [hunk 2]
```diff
@@ -2837,3 +2832,158 @@ def feature_store_name(self):
...
+            self.statistics_config = statistics_config
+        else:
+            self.primary_key = primary_key
+            self.statistics_config = statistics_config
+            self._features = features
+
+        self._href = href
+
+    def _save(self):
+        """Persist the metadata for this external feature group.
```
Comment (jimdowling) [inline python/hsfs/feature_group.py:2923]: external feature group -> spine group

File: python/hsfs/feature_store.py [hunk 3]
```diff
@@ -903,6 +903,146 @@ def create_external_feature_group(
...
+                version in the feature store.
+            description: A string describing the contents of the external feature group to
+                improve discoverability for Data Scientists, defaults to empty string
+                `""`.
+            primary_key: A list of feature names to be used as primary key for the
+                feature group. This primary key can be a composite key of multiple
+                features and will be used as joining key, if not specified otherwise.
+                Defaults to empty list `[]`, and the feature group won't have any primary key.
+            event_time: Optionally, provide the name of the feature containing the event
+                time for the features in this feature group. If event_time is set
```
[the remark this answers is not in the document]
Comment (jimdowling) [reply python/hsfs/feature_store.py:1006]: If the event_time is not set, you should still be able to do PiT correct joins - just with the current time as the event time, right?

File: python/tests/core/test_feature_view_engine.py [hunk 4]
```diff
@@ -1977,7 +1977,7 @@ def test_check_feature_group_accessibility_get_type_python(self, mocker):
         # Assert
         assert (
             str(e_info.value)
-            == "Python kernel can only read from cached feature group. Please use `feature_view.create_training_data` instead."
+            == "Python kernel can only read from cached feature groups. When using external feature groups please use `feature_view.create_training_data` instead. If you are using spines, use a Spark Kernel."
```
Comment (jimdowling) [inline python/tests/core/test_feature_view_engine.py:1980]: Do you have to use a Spark Kernel for spines?

**Review by jimdowling — COMMENTED** [review]
Typos in docstrings

**Review by jimdowling — APPROVED** [review]

File: python/hsfs/feature_store.py [hunk 5]
```diff
@@ -903,6 +903,146 @@ def create_external_feature_group(
...
+
+        The spine can then be used to construct queries, with only one speciality:
+
+        !!! note
+            Spines can only be used on the left side of a feature join, as this is the base
+            set of entities for which features are to be fetched and the left side of the join
+            determines the event timestamps to compare against.
+
+        **If you want to use the query for a feature view to be used for online serving,
+        you can only select the label or target feature from the spine.**
```
Comment (SirOibaf) [inline python/hsfs/feature_store.py:961]: I guess you mean that if you are using the spine, then the above applies. But if you are not using the spine the label can come from any feature group.

File: python/hsfs/feature_store.py [hunk 6]
```diff
@@ -903,6 +903,146 @@ def create_external_feature_group(
...
+        For the online lookup, the label is not required, therefore it is important to only
+        select label from the left feature group, so that we don't need to provide a spine
+        for online serving.
+
+        These queries can then be used to create feature views. Since the dataframe contained in the
+        spine is not being materialized, every time you use a feature view created with spine to read data
+        you will have to provide a dataframe with the same structure again.
```
Comment (SirOibaf) [inline python/hsfs/feature_store.py:968]: Do we do any validation that the dataframe matches the existing spine columns?

File: python/tests/core/test_feature_view_engine.py (hunk 4 above)
Comment (SirOibaf) [reply python/tests/core/test_feature_view_engine.py:1980]: I don't think we can get it working with Hive. But we should be able to get it working with Flyingduck.

Comment (SirOibaf) [reply python/tests/core/test_feature_view_engine.py:1980]: https://arrow.apache.org/docs/python/generated/pyarrow.flight.FlightClient.html#pyarrow.flight.FlightClient.do_put

File: python/hsfs/feature_store.py (hunk 5 above)
Comment (moritzmeister) [reply python/hsfs/feature_store.py:961]: Spines can always only be on the left side of a join. If you want to do online serving -> you can only select the label.
Of course if you don't use a spine at all, you can use any column as label. but this docs deals only with spines, so I am not sure what you are asking to change?

File: python/hsfs/feature_store.py (hunk 6 above)
Comment (moritzmeister) [reply python/hsfs/feature_store.py:968]: currently no, I rely on the join to fail.

File: python/tests/core/test_feature_view_engine.py (hunk 4 above)
Comment (moritzmeister) [reply python/tests/core/test_feature_view_engine.py:1980]: only spark for now unfortunately

File: python/hsfs/feature_store.py [hunk 7]
```diff
@@ -903,6 +903,146 @@ def create_external_feature_group(
...
+                improve discoverability for Data Scientists, defaults to empty string
+                `""`.
+            primary_key: A list of feature names to be used as primary key for the
+                feature group. This primary key can be a composite key of multiple
+                features and will be used as joining key, if not specified otherwise.
+                Defaults to empty list `[]`, and the feature group won't have any primary key.
+            event_time: Optionally, provide the name of the feature containing the event
+                time for the features in this feature group. If event_time is set
+                the feature group can be used for point-in-time joins. Defaults to `None`.
+            features: Optionally, define the schema of the external feature group manually as a
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply python/hsfs/feature_store.py]: no, labels are a concept of the feature view. the features parameter is there, if you want to you can create the spine without passing a dataframe to infer the schema, but instead overwrite it here.

File: python/hsfs/feature_store.py [hunk 8]
```diff
@@ -903,6 +903,146 @@ def create_external_feature_group(
...
+                will create the feature group with incremented version from the last
+                version in the feature store.
+            description: A string describing the contents of the external feature group to
+                improve discoverability for Data Scientists, defaults to empty string
+                `""`.
+            primary_key: A list of feature names to be used as primary key for the
+                feature group. This primary key can be a composite key of multiple
+                features and will be used as joining key, if not specified otherwise.
+                Defaults to empty list `[]`, and the feature group won't have any primary key.
+            event_time: Optionally, provide the name of the feature containing the event
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply python/hsfs/feature_store.py:1005]: it's optional if you don't want to do PIT join at all, but you always have to provide the column in the dataframe.

File: python/hsfs/feature_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (moritzmeister) [reply python/hsfs/feature_store.py]: No, you will always have to provide the event time column, you will have to add a column with the current timestamp.

File: python/hsfs/feature_group.py (hunk 2 above)
Comment (moritzmeister) [reply python/hsfs/feature_group.py]: it's a private method, not to be used by users, but fixed anyway

File: python/hsfs/feature_store.py (hunk 6 above)
Comment (moritzmeister) [reply python/hsfs/feature_store.py:968]: ah wait, should be able to use 
```

            self._verify_schema_compatibility(
                feature_group.features, dataframe_features
```

**Review by kennethmhc — APPROVED** [review]
LGTM!
